### PR TITLE
Store python default type as PyTensorType instead of at::Type

### PR DESCRIPTION
**Pull request by li-roy**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #17530 Small clean up of aten_op&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14237931/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #17601 Store ScalarType and Backend instead of Type in TensorIterator&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14274754/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #17785 Remove Type::elementSizeInBytes&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14379074/)
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#17723 Store python default type as PyTensorType instead of at::Type**&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14351082/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #17786 Pass ScalarType separately from Type in python constructors&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14379075/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #17792 Remove Type::ScalarType()&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14381774/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #17603 Remove conversion operator from Type to TensorOptions&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D14276588/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #17787 Add ScalarType arg to Type::options()&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D14379076/)

Differential Revision: [D14351082](https://our.internmc.facebook.com/intern/diff/D14351082/)

### CODE REVIEW

**Comment by ezyang** [comment]
@gchanan has wondered if we shouldn't have another type representing "backend and scalar type" (which could serve the function that Type previously served.) I'm not sure if it's necessary, but when I see you mucking around with `pair<Backend, ScalarType>` it seems like something to consider doing. (I'm ambivalent because it's not altogether clear why such a concept is well founded, as opposed to just Backend or TensorTypeId). What do you think?

File: torch/csrc/tensor/python_tensor.cpp [hunk 1]
```diff
@@ -252,6 +237,17 @@ static void initialize_aten_types(std::vector<PyTensorType>& tensor_types) {
   }
 }
 
+static PyTensorType& get_tensor_type(THPDtype *dtype, THPLayout *layout, bool is_cuda) {
```
Comment (ezyang) [inline torch/csrc/tensor/python_tensor.cpp]: any particular reason for this reordering? (did anything change?)

**Review by ezyang — APPROVED** [review]

**Comment by li-roy** [comment]
IMO the new type is unnecessary. The long term goal is to de-couple as much as possible anyways.

File: torch/csrc/tensor/python_tensor.cpp (hunk 1 above)
Comment (li-roy) [reply torch/csrc/tensor/python_tensor.cpp]: calling it in initialize_python_bindings now

**Review by gchanan — CHANGES_REQUESTED** [review]
Unless I'm misunderstanding something, this PR seems backwards to me.

We have a useful type for storing realized options of a Tensor -- `at::Type`.  Okay, it also does other things that we don't like, such as being used for (internal) dispatch, but that's relatively easy to fix -- use a different type for internal dispatch.

Instead, we are changing the interfaces to return a type I never want -- `PyTensorType`.

How feasible is it to keep `at::Type` is a minimal "holding" type, and introduce new dispatch types?

**Pull request closed without merging**